src/main.rs

```diff
@@ -1,3 +1,207 @@
-fn main() {
+use chrono::NaiveDateTime;
+use clap::Parser;
+use serde::{Deserialize, Serialize};
 
+use rayon::{prelude::*, current_num_threads};
+
+use std::borrow::Cow;
+use std::collections::HashMap;
+use std::hash::Hash;
+use std::path::PathBuf;
+use std::{fs, io};
+
+fn main() -> io::Result<()> {
+    let cli = Cli::parse();
+
+    if let Some(jobs) = cli.jobs {
+        rayon::ThreadPoolBuilder::new().num_threads(jobs).build_global().unwrap();
+    }
+
+    let content = fs::read_to_string(cli.file)?;
+
+    let chat: Chat = serde_json::from_str(&content).unwrap();
+    let stat: ChatStatistics = ChatStatistics::gather(&chat);
+
+    let file = fs::File::create(cli.output)?;
+    serde_json::to_writer_pretty(file, &stat)?;
+
+    Ok(())
+}
+
+#[derive(Debug, Parser)]
+struct Cli {
+    #[arg(long, short)]
+    file: PathBuf,
+    #[arg(long, short, default_value = "out.json")]
+    output: PathBuf,
+    #[arg(long, short)]
+    jobs: Option<usize>
+}
+
+#[derive(Debug, Serialize, Deserialize, Eq, PartialEq, Hash, Clone)]
+struct Person<'a>(Cow<'a, str>);
+
+#[derive(Debug, Serialize, Deserialize, Eq, PartialEq, Hash, Clone)]
+struct Token<'a>(Cow<'a, str>);
+
+impl<'a> From<&'a str> for Token<'a> {
+    fn from(value: &'a str) -> Self {
+        Token(value.into())
+    }
+}
+impl<'a> From<String> for Token<'a> {
+    fn from(value: String) -> Self {
+        Token(value.into())
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Chat<'a> {
+    name: String,
+    #[serde(rename = "type")]
+    chat_type: ChatType,
+    id: u128,
```

Comment:
It's a good practice to store that kind of type in e.g. `ChatId(u128)`, however, it's up to you.